# Dataframes
---
This is an introduction to reading, manipulating, and writing tabular data in python, focussing on the dataframe - the most common way to handle tables.

By the end you will:
- Learn how to read (and write) in excel and csv formats
- Inspect dataframes
- Subset dataframes

## Import packages

The dataframe is actually not part of python, it belongs to a python package called pandas.

We have to make our notebook aware of any packages we want to use by 'importing' them

In [26]:
import os # operating system: for interacting with files
import pandas as pd # key tabular data package in python
import numpy as np # important package for mathematics in python

## Reading data into dataframes

In this example our data is stored as a CSV file (Comma-Separated Values).

This is a simple way to store and organize data in a plain text format.

It’s like a spreadsheet but without all the extra features of software like Excel or Google Sheets.

How a CSV File Works:
 - It contains rows and columns, just like a table.
 - Each row represents a record (or an entry).
 - Each column contains a specific type of information (like a name, date, or price).
 - The values in each row are separated by commas (,), which is why it’s called Comma-Separated Values.

These data are from: _Machine learning has been used to predict patient survival based on key clinical parameters_ [Chicco & Jurman, 2020](https://doi.org/10.1186/s12911-020-1023-5).

We read in the csv file into a **pandas dataframe** called hf_prediction_data.

In [3]:
input_file_path = "./data/heart_failure_clinical_records_dataset.csv" # define the relative path to the file
hf_prediction_data = pd.read_csv(input_file_path) # read in the csv file using pandas built-in csv reading function

## Manipulating dataframes

In [4]:
# We can inspect the data
hf_prediction_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 299 entries, 0 to 298
Data columns (total 13 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   age                       299 non-null    float64
 1   anaemia                   299 non-null    int64  
 2   creatinine_phosphokinase  299 non-null    int64  
 3   diabetes                  299 non-null    int64  
 4   ejection_fraction         299 non-null    int64  
 5   high_blood_pressure       299 non-null    int64  
 6   platelets                 299 non-null    float64
 7   serum_creatinine          299 non-null    float64
 8   serum_sodium              299 non-null    int64  
 9   sex                       299 non-null    int64  
 10  smoking                   299 non-null    int64  
 11  time                      299 non-null    int64  
 12  DEATH_EVENT               299 non-null    int64  
dtypes: float64(3), int64(10)
memory usage: 30.5 KB


We can see the table has 12 columns and 299 rows. The Dtype indicates the data are either integers (int64) or non-integer numbers (float64)

In [28]:
hf_prediction_data.head() # the head() method shows us the top

age  anaemia  creatinine_phosphokinase  diabetes  ejection_fraction  \
0  75.0        0                       582         0                 20   
1  55.0        0                      7861         0                 38   
2  65.0        0                       146         0                 20   
3  50.0        1                       111         0                 20   
4  65.0        1                       160         1                 20   

   high_blood_pressure  platelets  serum_creatinine  serum_sodium  sex  \
0                    1  265000.00               1.9           130    1   
1                    0  263358.03               1.1           136    1   
2                    0  162000.00               1.3           129    1   
3                    0  210000.00               1.9           137    1   
4                    0  327000.00               2.7           116    0   

   smoking  time  DEATH_EVENT  
0        0     4            1  
1        0     6            1  
2        1     7            1  
3        0     7            1  
4        0     8            1

In [6]:
hf_prediction_data.tail() # the tail() method shows us the bottom

age  anaemia  creatinine_phosphokinase  diabetes  ejection_fraction  \
294  62.0        0                        61         1                 38   
295  55.0        0                      1820         0                 38   
296  45.0        0                      2060         1                 60   
297  45.0        0                      2413         0                 38   
298  50.0        0                       196         0                 45   

     high_blood_pressure  platelets  serum_creatinine  serum_sodium  sex  \
294                    1   155000.0               1.1           143    1   
295                    0   270000.0               1.2           139    0   
296                    0   742000.0               0.8           138    0   
297                    0   140000.0               1.4           140    1   
298                    0   395000.0               1.6           136    1   

     smoking  time  DEATH_EVENT  
294        1   270            0  
295        0   271            0  
296        0   278            0  
297        1   280            0  
298        1   285            0

In [7]:
hf_prediction_data[['age', 'ejection_fraction']] # This is how to select particular columns by name

age  ejection_fraction
0    75.0                 20
1    55.0                 38
2    65.0                 20
3    50.0                 20
4    65.0                 20
..    ...                ...
294  62.0                 38
295  55.0                 38
296  45.0                 60
297  45.0                 38
298  50.0                 45

[299 rows x 2 columns]

In [8]:
# Subset the data by a condition, e.g. age above 60
hf_prediction_data[hf_prediction_data['age'] > 60]

age  anaemia  creatinine_phosphokinase  diabetes  ejection_fraction  \
0    75.0        0                       582         0                 20   
2    65.0        0                       146         0                 20   
4    65.0        1                       160         1                 20   
5    90.0        1                        47         0                 40   
6    75.0        1                       246         0                 15   
..    ...      ...                       ...       ...                ...   
283  65.0        0                      1688         0                 38   
288  65.0        0                       892         1                 35   
289  90.0        1                       337         0                 38   
293  63.0        1                       103         1                 35   
294  62.0        0                        61         1                 38   

     high_blood_pressure  platelets  serum_creatinine  serum_sodium  sex  \
0                      1  265000.00               1.9           130    1   
2                      0  162000.00               1.3           129    1   
4                      0  327000.00               2.7           116    0   
5                      1  204000.00               2.1           132    1   
6                      0  127000.00               1.2           137    1   
..                   ...        ...               ...           ...  ...   
283                    0  263358.03               1.1           138    1   
288                    0  263358.03               1.1           142    0   
289                    0  390000.00               0.9           144    0   
293                    0  179000.00               0.9           136    1   
294                    1  155000.00               1.1           143    1   

     smoking  time  DEATH_EVENT  
0          0     4            1  
2          1     7            1  
4          0     8            1  
5          1     8            1  
6          0    10            1  
..       ...   ...          ...  
283        1   250            0  
288        0   256            0  
289        0   256            0  
293        1   270            0  
294        1   270            0  

[137 rows x 13 columns]

In [29]:
# Subset by 2 conditions: Find non-smokers (smoking == 0) with diabetes (diabetes == 1)
hf_prediction_data[
    (hf_prediction_data['smoking'] == 0) & # condition 1, the '&' means both must be satisfied
    (hf_prediction_data['diabetes'] == 1) # condition 2
]

age  anaemia  creatinine_phosphokinase  diabetes  ejection_fraction  \
4    65.0        1                       160         1                 20   
19   48.0        1                       582         1                 55   
21   65.0        1                       128         1                 30   
23   53.0        0                        63         1                 60   
24   75.0        0                       582         1                 30   
..    ...      ...                       ...       ...                ...   
285  55.0        1                       170         1                 40   
287  45.0        0                       582         1                 55   
288  65.0        0                       892         1                 35   
290  45.0        0                       615         1                 55   
296  45.0        0                      2060         1                 60   

     high_blood_pressure  platelets  serum_creatinine  serum_sodium  sex  \
4                      0  327000.00              2.70           116    0   
19                     0   87000.00              1.90           121    0   
21                     1  297000.00              1.60           136    0   
23                     0  368000.00              0.80           135    1   
24                     1  263358.03              1.83           134    0   
..                   ...        ...               ...           ...  ...   
285                    0  336000.00              1.20           135    1   
287                    0  543000.00              1.00           132    0   
288                    0  263358.03              1.10           142    0   
290                    0  222000.00              0.80           141    0   
296                    0  742000.00              0.80           138    0   

     smoking  time  DEATH_EVENT  
4          0     8            1  
19         0    15            1  
21         0    20            1  
23         0    22            0  
24         0    23            1  
..       ...   ...          ...  
285        0   250            0  
287        0   250            0  
288        0   256            0  
290        0   257            0  
296        0   278            0  

[95 rows x 13 columns]

In [10]:
# Get summary statistics
hf_prediction_data.describe()

age     anaemia  creatinine_phosphokinase    diabetes  \
count  299.000000  299.000000                299.000000  299.000000   
mean    60.833893    0.431438                581.839465    0.418060   
std     11.894809    0.496107                970.287881    0.494067   
min     40.000000    0.000000                 23.000000    0.000000   
25%     51.000000    0.000000                116.500000    0.000000   
50%     60.000000    0.000000                250.000000    0.000000   
75%     70.000000    1.000000                582.000000    1.000000   
max     95.000000    1.000000               7861.000000    1.000000   

       ejection_fraction  high_blood_pressure      platelets  \
count         299.000000           299.000000     299.000000   
mean           38.083612             0.351171  263358.029264   
std            11.834841             0.478136   97804.236869   
min            14.000000             0.000000   25100.000000   
25%            30.000000             0.000000  212500.000000   
50%            38.000000             0.000000  262000.000000   
75%            45.000000             1.000000  303500.000000   
max            80.000000             1.000000  850000.000000   

       serum_creatinine  serum_sodium         sex    smoking        time  \
count         299.00000    299.000000  299.000000  299.00000  299.000000   
mean            1.39388    136.625418    0.648829    0.32107  130.260870   
std             1.03451      4.412477    0.478136    0.46767   77.614208   
min             0.50000    113.000000    0.000000    0.00000    4.000000   
25%             0.90000    134.000000    0.000000    0.00000   73.000000   
50%             1.10000    137.000000    1.000000    0.00000  115.000000   
75%             1.40000    140.000000    1.000000    1.00000  203.000000   
max             9.40000    148.000000    1.000000    1.00000  285.000000   

       DEATH_EVENT  
count    299.00000  
mean       0.32107  
std        0.46767  
min        0.00000  
25%        0.00000  
50%        0.00000  
75%        1.00000  
max        1.00000

In [11]:
# Sort the table by values in one column
hf_prediction_data.sort_values(by='serum_creatinine', ascending=False)

age  anaemia  creatinine_phosphokinase  diabetes  ejection_fraction  \
9    80.0        1                       123         0                 35   
217  54.0        1                       427         0                 70   
52   60.0        0                      3964         1                 62   
131  60.0        1                      1082         1                 45   
28   58.0        1                        60         0                 38   
..    ...      ...                       ...       ...                ...   
186  50.0        0                       582         0                 50   
47   60.0        0                       582         1                 38   
57   60.0        1                       607         0                 40   
78   75.0        1                       203         1                 38   
250  50.0        0                      2522         0                 30   

     high_blood_pressure  platelets  serum_creatinine  serum_sodium  sex  \
9                      1  388000.00               9.4           133    1   
217                    1  151000.00               9.0           137    0   
52                     0  263358.03               6.8           146    0   
131                    0  250000.00               6.1           131    1   
28                     0  153000.00               5.8           134    1   
..                   ...        ...               ...           ...  ...   
186                    0  153000.00               0.6           134    0   
47                     1  451000.00               0.6           138    1   
57                     0  216000.00               0.6           138    1   
78                     1  283000.00               0.6           131    1   
250                    1  404000.00               0.5           139    0   

     smoking  time  DEATH_EVENT  
9          1    10            1  
217        0   196            1  
52         0    43            1  
131        0   107            0  
28         0    26            1  
..       ...   ...          ...  
186        0   172            1  
47         1    40            1  
57         1    54            0  
78         1    74            0  
250        0   214            0  

[299 rows x 13 columns]

In [12]:
# Group data by a categorical variable: e.g. to find the mean age by survival status
hf_prediction_data.groupby('DEATH_EVENT')['age'].mean()

DEATH_EVENT
0    58.761906
1    65.215281
Name: age, dtype: float64

In [13]:
# Drop a column we no longer need. Note how we 'over-write' the dataframe with the new version which has had the 'time' column removed
hf_prediction_data_time_removed = hf_prediction_data.drop(columns=['time'])
hf_prediction_data_time_removed

age  anaemia  creatinine_phosphokinase  diabetes  ejection_fraction  \
0    75.0        0                       582         0                 20   
1    55.0        0                      7861         0                 38   
2    65.0        0                       146         0                 20   
3    50.0        1                       111         0                 20   
4    65.0        1                       160         1                 20   
..    ...      ...                       ...       ...                ...   
294  62.0        0                        61         1                 38   
295  55.0        0                      1820         0                 38   
296  45.0        0                      2060         1                 60   
297  45.0        0                      2413         0                 38   
298  50.0        0                       196         0                 45   

     high_blood_pressure  platelets  serum_creatinine  serum_sodium  sex  \
0                      1  265000.00               1.9           130    1   
1                      0  263358.03               1.1           136    1   
2                      0  162000.00               1.3           129    1   
3                      0  210000.00               1.9           137    1   
4                      0  327000.00               2.7           116    0   
..                   ...        ...               ...           ...  ...   
294                    1  155000.00               1.1           143    1   
295                    0  270000.00               1.2           139    0   
296                    0  742000.00               0.8           138    0   
297                    0  140000.00               1.4           140    1   
298                    0  395000.00               1.6           136    1   

     smoking  DEATH_EVENT  
0          0            1  
1          0            1  
2          1            1  
3          0            1  
4          0            1  
..       ...          ...  
294        1            0  
295        0            0  
296        0            0  
297        1            0  
298        1            0  

[299 rows x 12 columns]

## Saving dataframes

In [14]:
# Save the file
output_file_path = "./data/heart_failure_clinical_records_dataset_edited.csv"
hf_prediction_data_time_removed.to_csv(output_file_path)